### PR TITLE
Compute global gradient norm according to 'enable_grad_norm_clip'

### PR DESCRIPTION
**Pull request by SherlockNoMad**
**Description**: As title

**Motivation and Context**
Global gradient norm should be computed for fp32 runs.

### CODE REVIEW

File: orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc
Comment (wschin) [inline orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc:342]: [nit] These two if-condition's can be merged into one if-block.

File: orttraining/orttraining/core/graph/adasum_optimizer_graph_builder.cc
Comment (wschin) [inline orttraining/orttraining/core/graph/adasum_optimizer_graph_builder.cc:146]: [nit] It seems we have too many if-else blocks, but I don't have a quick solution for this PR.

**Review by wschin — APPROVED** [review]

File: orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc
Comment (wschin) [inline orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc:169]: Why not just use `enable_grad_norm_clip` to control everything about gradient norm? We can set `enable_grad_norm_clip` in ConfigureForTraining based on precision mode.

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (wschin) [inline orttraining/orttraining/core/graph/training_op_defs.cc:1742]: >PassThrough [](start = 31, length = 11)

It looks like `Identity`. If so, why do we create the same op again?

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (wschin) [inline orttraining/orttraining/core/graph/optimizer_graph_builder.cc:481]: I think this should be controlled by one single flag `check_gradient_magnitude`. This flag can be determined in ConfigureForTraining.

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (wschin) [inline orttraining/orttraining/core/graph/optimizer_graph_builder.cc:487]: >global_grad_norm_argdef.name.empty() [](start = 6, length = 36)

Use `graph.GetNodeArg(global_grad_norm_argdef.name) != null` to determine if it exists.

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (wschin) [inline orttraining/orttraining/core/graph/optimizer_graph_builder.cc:489]: This is tricky. The inputs of `ReduceAllL2` should also be the outputs of `PassThrough`; it means we want to compute the gradient norm from non-updated tensor values. In addition, adding `PassThrough` may increase peak memory. It's better to check the memory consumption using Wei's tool before merging this.

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:487]: They are equivalent. 
At this point, then change is still in GraphDefs, and hasn't been applied on the graph.

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:489]: WeightUpdate happens after the RedcueAllL2 node, so we won't run into the case you mentioned here. 

So far from what we have seen, peak memory is reached around end of FW. PassThrough is guarding the end of BW, so the memory is probably not a concern here. 
In addition, in other training configs like distributed training and fp16 training, we are already having a barrier node (AllReduce, ReduceAllL2). So introducing the PassThrough here can't make things worse.

File: orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc:169]: In the case of fp16 run, user can still set enable_grad_norm_clip to false and we should not use gradient norm in the optimizer computation. 

So we can't override enable_grad_norm_clip for this purpose.

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:481]: I think this is a temporary condition as BF16 is not ready.

Comment (wschin) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:481]: I understand. It's a temporal condition which will be removed. If we determine `check_gradient_mmagnitude` in ConfigureForTraining and pass `check_gradient_mmagnitude` into this place, this file will have no change even when BF16 is done. If we spread such a condition around this file, it will be error-prone when we remove it.

---
In reply to: [520932982](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520932982) [](ancestors = 520932982)

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (wschin) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:487]: If there is an official way of checking if `ArgDef` exists, we should only use it instead of having several alternative ways (it's like Python v.s. Perl and the current winner is Python). After reading the document of `Graph`, I believe `GetNodeArg` is the right way.

---
In reply to: [520921889](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520921889) [](ancestors = 520921889)

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (wschin) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:489]: I agree that connecting PassThrough with ReduceAllL2 doesn't change the math. My intention was making semantic clearer.

I tried WaitEvent for BERT (not sure if it's large or tiny), which is also a kind of Barrier, and adding WaitEvent significantly increases peak memory usage.

---
In reply to: [520924873](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520924873) [](ancestors = 520924873)

File: orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc
Comment (wschin) [reply orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc:169]: I should write it clearer. Inside this function, a single boolean flag is enough for controlling if gradient norm is computed. You can move
```
my_flag = opt_graph_config_.enable_grad_norm_clip ||
      (opt_graph_config_.use_mixed_precision &&
       opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16
```
outside this function and pass `my_flag` into this function.

---
In reply to: [520930473](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520930473) [](ancestors = 520930473)

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:487]: OK. I added a Exist() function to the ArgDef class.

File: orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/allreduce_optimizer_graph_builder.cc:169]: done

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc:481]: refactored.

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (wschin) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc]: The signature of the new `Exist` better to be `Exist(const Graph& graph)` because graph is the scope of declaration. Inside the new `Exist`, there could be
```
bool ArgDef::ExistIn(const Graph& graph) {
  return graph.GetNodeArg(global_grad_norm_argdef.name) != null;
}
```

---
In reply to: [520966059](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520966059) [](ancestors = 520966059)

**Review by wschin — APPROVED** [review]
:shipit:

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/optimizer_graph_builder.cc]: discussed offline. the nodearg hasn't been added to graph at this point.

**Review by liqunfu — APPROVED** [review]

**Pull request merged**
